src/planner/binder/tableref/bind_table_function.cpp

```diff
@@ -23,43 +23,85 @@
 
 namespace duckdb {
 
-static bool IsTableInTableOutFunction(TableFunctionCatalogEntry &table_function) {
-	auto fun = table_function.functions.GetFunctionByOffset(0);
-	return table_function.functions.Size() == 1 && fun.arguments.size() == 1 &&
-	       fun.arguments[0].id() == LogicalTypeId::TABLE;
+enum class TableFunctionBindType { STANDARD_TABLE_FUNCTION, TABLE_IN_OUT_FUNCTION, TABLE_PARAMETER_FUNCTION };
+
+static TableFunctionBindType GetTableFunctionBindType(TableFunctionCatalogEntry &table_function,
+                                                      vector<unique_ptr<ParsedExpression>> &expressions) {
+	// first check if all expressions are scalar
+	// if they are we always bind as a standard table function
+	bool all_scalar = true;
+	for (auto &expr : expressions) {
+		if (!expr->IsScalar()) {
+			all_scalar = false;
+			break;
+		}
+	}
+	if (all_scalar) {
+		return TableFunctionBindType::STANDARD_TABLE_FUNCTION;
+	}
+	// if we have non-scalar parameters - we need to look at the function definition to decide how to bind
+	// if a function does not have an in_out_function defined, we need to bind as a standard table function regardless
+	bool has_in_out_function = false;
+	bool has_standard_table_function = false;
+	bool has_table_parameter = false;
+	for (idx_t function_idx = 0; function_idx < table_function.functions.Size(); function_idx++) {
+		const auto &function = table_function.functions.GetFunctionReferenceByOffset(function_idx);
+		for (auto &arg : function.arguments) {
+			if (arg.id() == LogicalTypeId::TABLE) {
+				has_table_parameter = true;
+			}
+		}
+		if (function.in_out_function) {
+			has_in_out_function = true;
+		} else if (function.function || function.bind_replace) {
+			has_standard_table_function = true;
+		} else {
+			throw InternalException("Function \"%s\" has neither in_out_function nor function defined",
+			                        table_function.name);
+		}
+	}
+	if (has_table_parameter) {
+		if (table_function.functions.Size() != 1) {
+			throw InternalException(
+			    "Function \"%s\" has a TABLE parameter, and multiple function overloads - this is not supported",
+			    table_function.name);
+		}
+		return TableFunctionBindType::TABLE_PARAMETER_FUNCTION;
+	}
+	if (has_in_out_function && has_standard_table_function) {
+		throw InternalException("Function \"%s\" is both an in_out_function and a table function", table_function.name);
```

Comment:
Do we want to throw an InternalException here?
This could happen if an extension registers a new overload for a function

Reply:
I think it's fine to throw an internal exception if an extension screws up - if we had to guard against internal exceptions from extensions that would complicate system design quite a bit :)